owltorrent-network/src/main/java/edu/rice/owltorrent/network/TorrentManager.java

```diff
@@ -0,0 +1,199 @@
+ * peer connections and ask for missing file pieces.
+ *
+ * @author Lorraine Lyu, Max Yu
+ */
+@Log4j2(topic = "general")
+public class TorrentManager implements Runnable, AutoCloseable {
+
+  private static final int DEFAULT_BLOCK_NUM = 2;
+
+  private static final int BLOCK_NOT_STARTED = 0;
+  private static final int BLOCK_IN_PROGRESS = 1;
+  private static final int BLOCK_DONE = 2;
+
+  private final Set<Integer> completedPieces;
+  private final Map<Integer, PieceStatus> uncompletedPieces;
+  private final Set<Integer> notStartedPieces;
+
+  private final int totalPieces;
+
+  private final Map<Peer, PeerConnector> peers;
+  private final NetworkToStorageAdapter networkStorageAdapter;
+  @Getter private final Torrent torrent;
+
+  public TorrentManager(Torrent file, NetworkToStorageAdapter adapter) {
+    this.torrent = file;
+    this.networkStorageAdapter = adapter;
+    this.completedPieces = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    this.uncompletedPieces = new ConcurrentHashMap<>();
+    this.notStartedPieces = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    this.totalPieces = torrent.getPieces().size();
+
+    for (int idx = 0; idx < totalPieces; idx++) {
+      this.notStartedPieces.add(idx);
+    }
+
+    this.peers = new ConcurrentHashMap<>();
+
+    // initPeers();
+  }
+
+  private void initPeers(List<Peer> peerList) {
+    for (Peer peer : peerList) {
+      try {
+        PeerConnector connector =
+            SocketConnector.makeInitialConnection(peer, this, networkStorageAdapter);
+        addPeer(connector, peer);
+        // TODO: revise
+        connector.writeMessage(new PayloadlessMessage(PeerMessage.MessageType.INTERESTED));
+      } catch (IOException e) {
+        log.error(String.format("Error connecting peer id=%s", peer.getPeerID().toString()));
+      }
+    }
+  }
+
+  public void startDownloadingAsynchronously() {
+    // TODO: make thread pool scheduler
+    new Thread(this).start();
+  }
+
+  public float getProgressPercent() {
+    return completedPieces.size() * 1.0f / totalPieces;
+  }
+
+  // For handshake listener
+  // Later if get updated peerlist from tracker.
+  public void addPeer(PeerConnector connector, Peer peer) {
+    this.peers.put(peer, connector);
+  }
+
+  @Override
+  public void close() throws Exception {
+    for (var pair : peers.entrySet()) {
+      pair.getValue().close();
+    }
+  }
+
+  private void requestPieceFromPeer(Peer peer, int pieceNum, int blockNum, int blockSize) {
+    PeerConnector peerConnector = peers.get(peer);
+    if (peerConnector == null) {
+      return;
+    }
+
+    // TODO: fix int casting.
+    try {
+      peerConnector.writeMessage(
+          PieceActionMessage.makeRequestMessage(pieceNum, blockNum * blockSize, blockSize));
+    } catch (IOException e) {
+      log.error(e);
+    }
+  }
+
+  @Override
+  public void run() {
+    while (!uncompletedPieces.isEmpty()) {
+      // Request a missing piece from each Peer
```

Comment:
Maybe I'm missing something, but how do we ensure that the peers we are requesting from have the block we want? Also, when we are requesting, since there isn't a sleep or a check for how many blocks we are currently requesting, as soon as this gets run we're going to request every single block at once, which is problematic as well.

Reply:
For now we're just assuming we're connecting to a seeder

Reply:
Okay, we should add a TODO in regards to this though